registry/modules/auth-basic/controllers/authBasic.ts

```diff
@@ -50,12 +76,34 @@ export const createAuthBasicController = (userRepo: UserRepository) => {
       const isOk = await argon.verify(hashedPass, password);
 
       if (isOk) {
+        const refreshToken = await generateRefreshToken(user.userId);
         const accessToken = generateAccessToken(user.userId);
 
-        return { user, accessToken };
+        return { user, accessToken, refreshToken };
       }
 
       throw invalidLoginCredentials();
     },
+
+    async refreshToken({ token }: RefreshTokenSchema) {
+      const tokenData = await refreshTokenRepo.getToken(token);
+
+      if (!tokenData) throw forbiddenError();
+
+      const { userId, tokenFamily, active } = tokenData;
+
+      if (active) {
+        // Token is valid and hasn't been used yet
+        const newRefreshToken = await generateRefreshToken(userId, tokenFamily);
+        const accessToken = generateAccessToken(userId);
+
+        return { accessToken, refreshToken: newRefreshToken };
+      } else {
+        // Previously refreshed token used, invalidate all tokens in family
```

Comment:
What happens in the scenario where a refresh token was used but the new token was not successfully returned to the client? Then the client is left without any valid token?

```mermaid
sequenceDiagram

client->>server: /useRefreshToken
server ->> server: issue new token
server ->> server: invalidate old token
server-->>client: TCP connection breaks and new token does not reach client
client-->>client: Client now has no valid tokens at all?
```